enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala

```diff
@@ -752,4 +752,69 @@ return p""")
     // Then
     result.toList should equal(List(Map("a" -> n)))
   }
+
+  test("Should handle optional match with null parts and distinct without NullPointerException") {
+
+    graph.execute("CREATE (:X {eid: 1})")
+
+    val query =
+      """
+        |  MATCH (x:X)
+        |  WITH x
+        |  OPTIONAL MATCH (x)<-[:L]-(req)
+        |  WITH x, req
+        |  OPTIONAL MATCH (req)<-[:Links *2]-(y:X)
+        |  RETURN DISTINCT
+        |    x.eid,
+        |    req.eid,
+        |    y.eid
+      """.stripMargin
+
+    val result = executeWithAllPlannersAndCompatibilityMode(query)
+
+    result.toList should equal(List(Map("x.eid" -> 1, "req.eid" -> null, "y.eid" -> null)))
+  }
+
+  test("Should handle complex optional match with null parts and distinct without NullPointerException") {
+
+    graph.execute(
+      """
+        |CREATE (cust :DoorsObject {eid: 1})
+        |  CREATE (cust) <-[:Links]- (req :DoorsObject {eid: 1})
+        |  CREATE (reqM :DoorsFormalModule {eid: 1}) -[:Contains]-> (req)
+      """.stripMargin)
+
+    graph.execute(
+      """
+        |CREATE (cust :DoorsObject {eid: 2})
+        |  CREATE (cust) <-[:Links]- (req :DoorsObject {eid: 2})
+        |  CREATE (reqM :DoorsFormalModule {eid: 2}) -[:Contains]-> (req)
+        |  CREATE (req) <-[:Links]- (:Link) <-[:Links]- (tst :DoorsObject {eid: 2})
+        |  CREATE (tstM :DoorsFormalModule {eid: 2}) -[:Contains]-> (tst)
+      """.stripMargin)
+
+    val query =
+      """
+        |  MATCH (cust :DoorsObject)
+        |  WITH cust
+        |  OPTIONAL MATCH (cust)<-[:Links *1..]-(req :DoorsObject)<-[:Contains]-(reqM :DoorsFormalModule)
+        |  WITH DISTINCT cust, reqM, req
+        |  OPTIONAL MATCH (req)<-[:Links *2]-(tst :DoorsObject)<-[:Contains]-(tstM :DoorsFormalModule)
+        |  WITH cust, reqM, req, tstM, tst
+        |  RETURN DISTINCT
+        |    cust.eid,
+        |    reqM.eid,
+        |    tstM.eid
+      """.stripMargin
```

Comment:
Does this test trigger a bug which is not caught by the simpler test above?

Reply:
Not really, it's the same bug but this test is the code from the Github issue, while the simpler test is the simplest query we could found that generated the same error, so it's more for understanding.

Reply:
So do we need this test then, if it adds no information? The reason I'm fretting is that this test is sort of hard to understand to me at least: it has two quite complex setup queries, and a quite big test query, all with labels and types and var-length patterns. When this test fails, future me will have to parse through all of that to be able to know what could cause the failure. If it does not expose any known problem that the short test cannot expose, it is much easier to not have to read it =). 

I'm open to be convinced though!